### PR TITLE
Improvements of early exit for find-based algorithms (when we search for first/last matching data entry)

### PR DESCRIPTION
**Pull request by SergeyKopienko**
This PR based on idea of @danhoeflinger described at https://github.com/oneapi-src/oneDPL/pull/1689#discussion_r1672540484 from PR https://github.com/oneapi-src/oneDPL/pull/1689

In this PR, we implemented the early exit from search inside one sub-group in all sub-group items for find-based algorithms
(when we search for first/last matching data entry).

It's give us some significant performance boost.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 1]
```diff
@@ -1084,45 +1086,41 @@ struct __early_exit_find_or
 {
     _Pred __pred;
 
-    template <typename _NDItemId, typename _IterSize, typename _WgSize, typename _LocalFoundState, typename _BrickTag,
-              typename... _Ranges>
+    template <typename _NDItemId, typename _SrcDataSize, typename _DataPerWorkItemSize, typename _LocalFoundState,
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1089]: It looks like `_DataPerWorkItemSize` is outdated. Should we rename it to something like `_IterationDataSize` to match the parameter's name?

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 2]
```diff
@@ -1084,45 +1086,41 @@ struct __early_exit_find_or
...
+              typename _BrickTag, typename... _Ranges>
     void
-    operator()(const _NDItemId __item_id, const _IterSize __n_iter, const _WgSize __wg_size,
-               _LocalFoundState& __found_local, _BrickTag __brick_tag, _Ranges&&... __rngs) const
+    operator()(const _NDItemId __item_id, const _SrcDataSize __source_data_size,
+               const _DataPerWorkItemSize __iteration_data_size, _LocalFoundState& __found_local, _BrickTag __brick_tag,
+               _Ranges&&... __rngs) const
     {
+        // Calculate the amount of processing data in each work-item
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1096]: I'm in favor of calling this `items` or `iters` per work-item to match other places where we use this. I would also suggest rephrasing this comment since it's not the amount of data but the number of elements. Something like: `// Calculate the number of elements to be processed by each work-item`.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 3]
```diff
@@ -1084,45 +1086,41 @@ struct __early_exit_find_or
...
-        ::std::size_t __group_idx = __item_id.get_group(0);
-
-        // each work_item processes N_ELEMENTS with step SHIFT
-        ::std::size_t __leader = (__local_idx / __shift) * __shift;
-        ::std::size_t __init_index = __group_idx * __wg_size * __n_iter + __leader * __n_iter + __local_idx % __shift;
-
-        // if our "line" is out of work group size, reduce the line to the number of the rest elements
-        if (__wg_size - __leader < __shift)
-            __shift = __wg_size - __leader;
+        // Return the index on this item in the Kernel's execution range
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1106]:
```suggestion
        // Return the index of this item in the kernel's execution range
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 4]
```diff
@@ -1084,45 +1086,41 @@ struct __early_exit_find_or
...
+        const auto __item_global_linear_id = __item_id.get_global_linear_id();
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1107]: Nit: We could shorten this to something like `__global_idx` to improve readability.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 5]
```diff
@@ -1226,7 +1216,10 @@ __parallel_find_or(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPoli
                     _AtomicType __found_local = __init_value;
 
                     // 2. Find any element that satisfies pred
-                    __pred(__item_id, __n_iter, __wgroup_size, __found_local, __brick_tag, __rngs...);
+                    //  - after this call __found_local may still have initial value:
+                    //    1) if no element satisfies pred;
+                    //    2) early exit from sub-group occurred: in this case the state of __found_local will updated on the next group operation
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1221]:
```suggestion
                    //    2) early exit from sub-group occurred: in this case the state of __found_local will updated in the next group operation (3)
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h (hunk 1 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h (hunk 2 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Comment fixed.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed: the variable `__data_per_work_item` has been renamed to __iters_per_work_item`

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h (hunk 5 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1221]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h (hunk 4 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1107]: May be `__global_id` ?

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 6]
```diff
@@ -1084,45 +1086,41 @@ struct __early_exit_find_or
...
+        for (_DataPerWorkItemSize __i = 0; !__something_was_found && __i < __iters_per_work_item; ++__i)
         {
-            _IterSize __current_iter = __i;
+            auto __local_src_data_idx = __i;
             if constexpr (__is_backward_tag(__brick_tag))
-                __current_iter = __n_iter - 1 - __i;
-
-            const auto __shifted_idx = __init_index + __current_iter * __shift;
+            {
+                assert(__iters_per_work_item >= __i + 1);
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1115]: Do we really need this assertion? `__i` should always be smaller than `__iters_per_work_item` based on the for loop. Leaving this would slow down the execution significantly when compiled with assertions.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 7]
```diff
@@ -1084,45 +1086,41 @@ struct __early_exit_find_or
...
-
-        // if our "line" is out of work group size, reduce the line to the number of the rest elements
-        if (__wg_size - __leader < __shift)
-            __shift = __wg_size - __leader;
+        // Return the index of this item in the kernel's execution range
+        const auto __global_id = __item_id.get_global_linear_id();
 
         bool __something_was_found = false;
-        for (_IterSize __i = 0; !__something_was_found && __i < __n_iter; ++__i)
+        for (_DataPerWorkItemSize __i = 0; !__something_was_found && __i < __iters_per_work_item; ++__i)
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1110]:
```suggestion
        for (_IterationDataSize__i = 0; !__something_was_found && __i < __iters_per_work_item; ++__i)
```

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed by another way

**Review by julianmi — DISMISSED** [review]
LGTM. Please wait for green CI and for @danhoeflinger's review since he proposed this approach.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h [hunk 8]
```diff
@@ -1084,45 +1086,40 @@ struct __early_exit_find_or
...
-               _LocalFoundState& __found_local, _BrickTag __brick_tag, _Ranges&&... __rngs) const
+    operator()(const _NDItemId __item_id, const _SrcDataSize __source_data_size,
+               const _IterationDataSize __iteration_data_size, _LocalFoundState& __found_local, _BrickTag __brick_tag,
+               _Ranges&&... __rngs) const
     {
+        // Calculate the number of elements to be processed by each work-item.
+        using _Iteration =
+            decltype(oneapi::dpl::__internal::__dpl_ceiling_div(__source_data_size, __iteration_data_size));
+        const _Iteration __iters_per_work_item =
+            oneapi::dpl::__internal::__dpl_ceiling_div(__source_data_size, __iteration_data_size);
```
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1100]: Can we just use `auto` and `std::decltype(__iters_per_work_item)` to do this without repeating the expression?

Also, do we expect this type to be something other than `_SrcDataSize`?  It seems like that is what it will be.

**Review by danhoeflinger — COMMENTED** [review]
Other than my minor question about types, this LGTM, and does a good job of simplifying the access pattern.
Thanks for pursuing this!

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h (hunk 8 above)
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: Fixed

**Comment by SergeyKopienko** [comment]
> Other than my minor question about types, this LGTM, and does a good job of simplifying the access pattern. Thanks for pursuing this!

It was done together with you, @danhoeflinger, and @julianmi 
Thanks a lot for your ideas and help!

**Review by julianmi — APPROVED** [review]

**Review by danhoeflinger — APPROVED** [review]
LGTM
